### PR TITLE
refactor 

### PR DESCRIPTION
**Pull request by a3828162**
- [x] consumer finish
- [x] api_{SVC}.go
- [x] processor
- [x] server
- [x] router

### CODE REVIEW

**Comment by ianchen0119** [comment]
Hi @a3828162 

Please help to resolve the conflicts in `pkg/service/init.go`.
Tks!

File: pkg/service/init.go [hunk 1]
```diff
@@ -158,19 +204,66 @@ func (a *UdmApp) Start(tlsKeyLogPath string) {
...
-		logger.InitLog.Errorf("Deregister NF instance Failed Problem[%+v]", problemDetails)
+		logger.MainLog.Errorf("Deregister NF instance Failed Problem[%+v]", problemDetails)
 	} else if err != nil {
-		logger.InitLog.Errorf("Deregister NF instance Error[%+v]", err)
+		logger.MainLog.Errorf("Deregister NF instance Error[%+v]", err)
 	} else {
-		logger.InitLog.Infof("Deregister from NRF successfully")
+		logger.MainLog.Infof("Deregister from NRF successfully")
+	}
+	logger.MainLog.Infof("CHF SBI Server terminated")
```
Comment (yccodr) [inline pkg/service/init.go:224]: @a3828162 Hello?

File: cmd/main.go [hunk 2]
```diff
@@ -52,18 +55,29 @@ func action(cliCtx *cli.Context) error {
...
+
+	udm, err := service.NewApp(ctx, cfg, tlsKeyLogPath)
 	if err != nil {
 		return err
 	}
 	UDM = udm
 
-	udm.Start(tlsKeyLogPath)
+	udm.Start()
+	UDM.WaitRoutineStopped()
```
Comment (ianchen0119) [inline cmd/main.go:80]: `udm.WaitRoutineStopped()`

File: internal/sbi/consumer/nf_management_test.go [hunk 3]
```diff
@@ -20,6 +19,6 @@ func TestSendRegisterNFInstance(t *testing.T) {
 		Reply(200).
 		JSON(map[string]string{})
 
-	_, _, err := SendRegisterNFInstance("http://127.0.0.10:8000", "1", models.NfProfile{})
-	require.NoError(t, err)
+	//_, _, err := SendRegisterNFInstance("http://127.0.0.10:8000", "1", models.NfProfile{})
```
Comment (ianchen0119) [inline internal/sbi/consumer/nf_management_test.go:22]: Why did you change the unit test?

File: internal/sbi/eventexposure/api_create_ee_subscription.go [hunk 4]
```diff
@@ -18,12 +18,11 @@ import (
 	"github.com/free5gc/openapi/models"
 	"github.com/free5gc/udm/internal/logger"
 	"github.com/free5gc/udm/internal/sbi/producer"
-	"github.com/free5gc/util/httpwrapper"
 )
 
 // HTTPCreateEeSubscription - Subscribe
 func HTTPCreateEeSubscription(c *gin.Context) {
-	var eeSubscriptionReq models.EeSubscription
+	var eesubscription models.EeSubscription
```
[the remark this answers is not in the document]
Comment (ianchen0119) [reply internal/sbi/eventexposure/api_create_ee_subscription.go:25]: Please also check on these folders:
image

File: cmd/main.go [hunk 5]
```diff
@@ -52,18 +55,29 @@ func action(cliCtx *cli.Context) error {
...
 		return err
 	}
 	factory.UdmConfig = cfg
-	udm, err := service.NewApp(cfg)
+
+	udm, err := service.NewApp(ctx, cfg, tlsKeyLogPath)
 	if err != nil {
 		return err
 	}
 	UDM = udm
```
Comment (ianchen0119) [inline cmd/main.go:77]: This line is duplicated (please see `NewApp()`).

File: internal/sbi/processor/processor.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+type Processor struct {
+	ProcessorUdm
+	consumer *consumer.Consumer
+}
+
+type HandlerResponse struct {
+	Status  int
+	Headers map[string][]string
+	Body    interface{}
+}
```
Comment (andy89923) [inline internal/sbi/processor/processor.go:23]: I thinks we don't need this.

File: internal/sbi/router.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+			group.PUT(route.Pattern, route.HandlerFunc)
+		case "DELETE":
+			group.DELETE(route.Pattern, route.HandlerFunc)
+		case "PATCH":
+			group.PATCH(route.Pattern, route.HandlerFunc)
+		}
+	}
+}
+
+func AddSDMService(group *gin.RouterGroup, routes []Route) {
```
Comment (andy89923) [inline internal/sbi/router.go:40]: .

File: internal/sbi/router.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+			group.PUT(route.Pattern, route.HandlerFunc)
+		case "DELETE":
+			group.DELETE(route.Pattern, route.HandlerFunc)
+		case "PATCH":
+			group.PATCH(route.Pattern, route.HandlerFunc)
+		}
+	}
+}
+
+func AddUEAUService(group *gin.RouterGroup, routes []Route) {
```
Comment (andy89923) [inline internal/sbi/router.go:57]: .

File: internal/sbi/router.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	Pattern string
+	// HandlerFunc is the handler function of this route.
+	HandlerFunc gin.HandlerFunc
+}
+
+type RouteGroup interface {
+	AddService(engine *gin.Engine) *gin.RouterGroup
+}
+
+func AddService(group *gin.RouterGroup, routes []Route) {
```
Comment (andy89923) [inline internal/sbi/router.go:23]: .

File: internal/sbi/router.go (hunk 7 above)
Comment (andy89923) [inline internal/sbi/router.go:40]: .

File: internal/sbi/router.go (hunk 8 above)
Comment (andy89923) [inline internal/sbi/router.go:57]: .

File: pkg/service/init.go [hunk 10]
```diff
@@ -1,41 +1,80 @@
...
+	consumer  *consumer.Consumer
+	processor *processor.Processor
 }
 
-func NewApp(cfg *factory.Config) (*UdmApp, error) {
-	udm := &UdmApp{cfg: cfg}
+func GetApp() *UdmApp {
+	return UDM
+}
+
```
Comment (andy89923) [inline pkg/service/init.go:46]: remove

File: internal/sbi/consumer/nrf_service.go [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+	profile.Ipv4Addresses = append(profile.Ipv4Addresses, udmContext.RegisterIPv4)
+	services := []models.NfService{}
+	for _, nfService := range udmContext.NfService {
+		services = append(services, nfService)
+	}
+	if len(services) > 0 {
+		profile.NfServices = &services
+	}
+	profile.UdmInfo = &models.UdmInfo{
+		// Todo
```
Comment (ming-hsien) [inline internal/sbi/consumer/nrf_service.go:207]: I think we can delete the unnecessary code.

File: pkg/service/init.go [hunk 12]
```diff
@@ -1,41 +1,72 @@
...
 )
 
+var _ app.App = &UdmApp{}
+
 type UdmApp struct {
-	cfg    *factory.Config
+	app.App
+	consumer.ConsumerUdm
+	processor.ProcessorUdm
+	sbi.ServerUdm
```
Comment (andy89923) [inline pkg/service/init.go:27]: remove, we don't need this

File: internal/sbi/consumer/nrf_service.go [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+func (s *nnrfService) getNFManagementClient(uri string) *Nnrf_NFManagement.APIClient {
+	if uri == "" {
+		return nil
+	}
+	s.nfMngmntMu.RLock()
+	client, ok := s.nfMngmntClients[uri]
+	if ok {
+		defer s.nfMngmntMu.RUnlock()
```
Comment (andy89923) [inline internal/sbi/consumer/nrf_service.go:38]: no defer
```suggestion
		s.nfMngmntMu.RUnlock()
```

Comment (andy89923) [reply internal/sbi/consumer/nrf_service.go:38]: so as others

**Review by andy89923 — COMMENTED** [review]
Also, need to fix the Termination procedure; you could refer to the Discord message.

File: internal/sbi/server.go [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+			cfg.GetCertPemPath(),
+			cfg.GetCertKeyPath())
+	} else {
+		err = fmt.Errorf("No support this scheme[%s]", scheme)
+	}
+
+	if err != nil && err != http.ErrServerClosed {
+		logger.SBILog.Errorf("SBI server error: %v", err)
+	}
+	logger.SBILog.Warnf("SBI server (listen on %s) stopped", s.httpServer.Addr)
```
Comment (andy89923) [inline internal/sbi/server.go:105]: suggest log level to info

File: internal/sbi/server.go [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+
+func newRouter(s *Server) *gin.Engine {
+	router := logger_util.NewGinWithLogrus(logger.GinLog)
+
+	// EE
+	udmEERoutes := s.getEventExposureRoutes()
+	udmEEGroup := s.router.Group(factory.UdmEeResUriPrefix)
+	routerAuthorizationCheck := util.NewRouterAuthorizationCheck(models.ServiceName_NUDM_EE)
+	udmEEGroup.Use(func(c *gin.Context) {
+		routerAuthorizationCheck.Check(c, udm_context.GetSelf())
```
Comment (andy89923) [inline internal/sbi/server.go:149]:
```suggestion
		routerAuthorizationCheck.Check(c, s.Context())
```

File: internal/sbi/processor/callback.go
Comment (andy89923) [inline internal/sbi/processor/callback.go:1]: `callback` rename to `notifier`

File: internal/sbi/api_uecontextmanagement.go [hunk 16]
```diff
@@ -0,0 +1,522 @@
...
+		return
+	} else {
+		c.Status(http.StatusNoContent)
+		return
+	}
+}
+
+// DeregistrationSmsfNon3gppAccess - delete SMSF registration for non 3GPP access
+func (s *Server) HandleDeregistrationSmsfNon3gppAccess(c *gin.Context) {
+	c.JSON(http.StatusOK, gin.H{})
```
Comment (andy89923) [inline internal/sbi/api_uecontextmanagement.go:380]: This could also change to http.NotImplement

File: internal/sbi/server.go [hunk 17]
```diff
@@ -102,7 +102,7 @@ func (s *Server) startServer(wg *sync.WaitGroup) {
 	if err != nil && err != http.ErrServerClosed {
 		logger.SBILog.Errorf("SBI server error: %v", err)
 	}
-	logger.SBILog.Warnf("SBI server (listen on %s) stopped", s.httpServer.Addr)
+	logger.SBILog.Info("SBI server (listen on %s) stopped", s.httpServer.Addr)
```
Comment (ianchen0119) [inline internal/sbi/server.go]: Use `Infof` instead.

File: internal/sbi/processor/enent_exposure.go [hunk 18]
```diff
@@ -1,4 +1,4 @@
-package producer
```
Comment (ianchen0119) [inline internal/sbi/processor/enent_exposure.go:1]: `xxx_enent` -> `xxx_event.go`

**Review by ianchen0119 — CHANGES_REQUESTED** [review]
Please ensure the implementations of processor pkg follow our standard.

File: internal/sbi/processor/ue_context_management.go [hunk 19]
```diff
@@ -400,69 +425,76 @@ func (p *Processor) UpdateAmfNon3gppAccessProcedure(
...
 		}
 	}()
 
-	return nil
+	c.Status(http.StatusNoContent)
 }
 
 func (p *Processor) RegistrationSmfRegistrationsProcedure(
+	c *gin.Context,
 	request *models.SmfRegistration,
```
Comment (andy89923) [inline internal/sbi/processor/ue_context_management.go:490]: `request` rename to `smfRegistration`

File: internal/sbi/processor/ue_context_management.go [hunk 20]
```diff
@@ -502,11 +537,14 @@ func (p *Processor) RegistrationSmfRegistrationsProcedure(
...
+		c.Status(http.StatusNoContent)
 	} else {
-		header = make(http.Header)
+		header := make(http.Header)
 		udmUe, _ := udm_context.GetSelf().UdmUeFindBySupi(ueID)
 		header.Set("Location", udmUe.GetLocationURI(udm_context.LocationUriSmfRegistration))
-		return header, request, nil
+		for key, val := range header { // header response is optional
+			c.Header(key, val[0])
+		}
```
[the remark this answers is not in the document]
Comment (andy89923) [reply internal/sbi/processor/ue_context_management.go:547]: I think we will have "Location" only in the header

File: internal/sbi/api_parameterprovision.go [hunk 21]
```diff
@@ -53,21 +60,10 @@ func HTTPUpdate(c *gin.Context) {
 		return
 	}
 
-	req := httpwrapper.NewRequest(c.Request, ppDataReq)
-	req.Params["gspi"] = c.Params.ByName("gpsi")
+	gpsi := c.Params.ByName("gpsi")
```
Comment (andy89923) [inline internal/sbi/api_parameterprovision.go:63]:
```go
gpsi := c.Params.ByName("gpsi")
if gpsi == "" {
  problem := ...
  
  C.JSON()
  return
}
```

File: internal/sbi/api_subscriberdatamanagement.go [hunk 22]
```diff
@@ -0,0 +1,617 @@
...
+			Status: http.StatusBadRequest,
+			Cause:  "No get plmn-id",
+		}
+		return nil, problemDetails
+	}
+}
+
+// Info - Nudm_Sdm Info service operation
+func (s *Server) HandleInfo(c *gin.Context) {
+	c.JSON(http.StatusOK, gin.H{})
```
Comment (andy89923) [inline internal/sbi/api_subscriberdatamanagement.go:74]: NotImplement

File: internal/sbi/api_subscriberdatamanagement.go [hunk 23]
```diff
@@ -0,0 +1,617 @@
...
+}
+
+// PutUpuAck - Nudm_Sdm Info for UPU service operation
+func (s *Server) HandlePutUpuAck(c *gin.Context) {
+	c.JSON(http.StatusOK, gin.H{})
```
Comment (andy89923) [inline internal/sbi/api_subscriberdatamanagement.go:79]: NotImplement

File: internal/sbi/api_subscriberdatamanagement.go [hunk 24]
```diff
@@ -0,0 +1,617 @@
...
+			Title:  "Malformed request syntax",
+			Status: http.StatusBadRequest,
+			Detail: problemDetail,
+		}
+		logger.SdmLog.Errorln(problemDetail)
+		c.JSON(http.StatusBadRequest, rsp)
+		return
+	}
+
+	// step 1: log
```
Comment (andy89923) [inline internal/sbi/api_subscriberdatamanagement.go:183]: remove

Comment (andy89923) [reply internal/sbi/api_subscriberdatamanagement.go:183]: remove all unnecessary comment message

**Review by andy89923 — APPROVED** [review]

File: internal/sbi/processor/processor.go [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+
+	Consumer() *consumer.Consumer
+}
+
+type Processor struct {
+	ProcessorUdm
+	consumer *consumer.Consumer
+}
+
+func NewProcessor(udm ProcessorUdm) (*Processor, error) {
```
Comment (ianchen0119) [inline internal/sbi/processor/processor.go:19]: Please refer to https://github.com/free5gc/amf/pull/132/files

You don't need to implement the Consumer getter here.

**Review by ianchen0119 — COMMENTED** [review]
LGTM (except for my commented part)

**Pull request merged**